### PR TITLE
Lokole offline email, take 2

### PR DESCRIPTION
**Pull request by veeara282**
### Fixes Bug
N/A

### Description of changes proposed in this pull request.
Add a new role for the Lokole web app, offline email only.

### Smoke-tested in operating system.
Ubuntu, Travis CI

We have a staging server at http://iiab.eastus.cloudapp.azure.com/lokole/. Please test the following and comment about your results here:
- Account creation and deletion
- Sending and receiving emails (local only)

Known bugs: Lokole is not listed on the app portal at [/home/](http://iiab.eastus.cloudapp.azure.com/home/) on the staging server. This was because the task that registers it raised an out-of-memory error every time we tried to run it.

### Mention a team member for further information or comment using @ name
@holta This is a retake of #894.

### CODE REVIEW

File: roles/lokole/tasks/install.yml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  when: not lokole_enabled
+
+- name: Copy lokole httpd conf file
+  template:
+    src: lokole.conf.j2
+    dest: "/etc/{{ apache_config_dir }}/lokole.conf"
+
+- name: Enable httpd conf file if we are disabled (debuntu)
+  file:
+    src: /etc/{{ apache_config_dir }}/lokole.conf
```
Comment (holta) [inline roles/lokole/tasks/install.yml]:
```suggestion
    src: "/etc/{{ apache_config_dir }}/lokole.conf"
```

File: roles/lokole/tasks/install.yml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    enabled: no
+    state: stopped
+  when: not lokole_enabled
+
+- name: Copy lokole httpd conf file
+  template:
+    src: lokole.conf.j2
+    dest: "/etc/{{ apache_config_dir }}/lokole.conf"
+
+- name: Enable httpd conf file if we are disabled (debuntu)
```
Comment (holta) [inline roles/lokole/tasks/install.yml]:
```suggestion
- name: Enable httpd conf file if lokole_enabled (debuntu)
```

File: roles/lokole/tasks/install.yml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    path: /etc/apache2/sites-enabled/lokole.conf
+    state: absent
+  when: not lokole_enabled and is_debuntu
+
+- name: Restart Apache, so it picks up the new aliases
+  service:
+    name: "{{ apache_service }}"
+    state: restarted
+
+- name: Add 'lokole' to list of services at /etc/iiab/iiab.ini
```
Comment (holta) [inline roles/lokole/tasks/install.yml]:
```suggestion
- name: Add 'lokole' variable values to {{ iiab_ini_file }}
```

File: roles/lokole/tasks/install.yml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  ini_file:
+    dest: "{{ service_filelist }}"
```
Comment (holta) [inline roles/lokole/tasks/install.yml]:
```suggestion
    path: "{{ iiab_ini_file }}"
```

File: roles/lokole/defaults/main.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+---
+# Information needed to install Lokole
+lokole_version: "0.1.24"
+lokole_install_path: "{{ content_base }}/lokole"
+lokole_venv: "{{ lokole_install_path }}/venv"
+
+# Information needed to run Lokole
+lokole_user: lokole
+lokole_run_directory: /home/{{ lokole_user }}/state
```
Comment (holta) [inline roles/lokole/defaults/main.yml]:
```suggestion
lokole_run_directory: "/home/{{ lokole_user }}/state"
```

File: roles/lokole/defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+lokole_enabled: True
+
+lokole_url: /lokole
+lokole_full_url: "http://{{ iiab_hostname }}.{{ iiab_domain }}{{ lokole_url }}"
+
+lokole_domain_socket: "{{ lokole_run_directory }}/lokole_gunicorn.sock"
+
+# Global variables provided by setup-lokole.sh L157-166
+# https://github.com/ascoderu/opwen-webapp/blob/master/setup/setup-lokole.sh#L157
+opwen_server_locale: '{{ default_language }}'
```
Comment (holta) [inline roles/lokole/defaults/main.yml]:
```suggestion
opwen_server_locale: "{{ default_language }}"
```

File: roles/lokole/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+opwen_server_timezone: '{{ local_tz }}'
```
Comment (holta) [inline roles/lokole/defaults/main.yml]:
```suggestion
opwen_server_timezone: "{{ local_tz }}"
```

File: roles/lokole/tasks/install.yml [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    src: webapp.sh.j2
+    dest: "{{ lokole_run_directory }}/webapp.sh"
+    mode: a+x
+  tags:
+    - configure
+
+- name: Create 'lokole' service
+  template:
+    src: lokole.service.j2
+    dest: "/etc/systemd/system/lokole.service"
```
Comment (holta) [inline roles/lokole/tasks/install.yml]:
```suggestion
    dest: /etc/systemd/system/lokole.service
```

**Review by holta — COMMENTED** [review]
Suggestions for @aidan-fitz (some of which appear necessary!)

**Comment by veeara282** [comment]
@holta, how do I reinstall Lokole based on the changes you proposed and remove all of the other apps?

**Comment by holta** [comment]
> how do I reinstall Lokole

Does the [./runrole](https://github.com/iiab/iiab/blob/master/runrole) command hopefully help?
```
cd /opt/iiab/iiab
./runrole lokole
```
Similarly this can be very useful: _(of course slower than the above)_
```
cd /opt/iiab/iiab
./iiab-install --reinstall
```
> remove all of the other apps

I don't know a clean way to accomplish that.

A fresh IIAB install is safest using http://download.iiab.io/6.7 on a new VM &mdash; when it asks you "Edit /etc/iiab/local_vars.yml to customize your Internet-in-a-Box? [Y/n]" feel free to slim down your [/etc/iiab/local_vars.yml](http://wiki.laptop.org/go/IIAB/local_vars_min.yml) so it contains `lokole_install: True` and `lokole_enabled: True` with most other apps set to False.

**Comment by holta** [comment]
@aidan-fitz I successfully ran `cd /opt/iiab/iiab` and `./runrole lokole` to install Lokole on top of an existing IIAB 6.7/master on Ubuntu 18.04.

After this completed I tried to browse to http://box/lokole and http://192.168.0.121/lokole and http://10.8.0.66/lokole and (in each case) got a response like this:
```
Not Found

The requested URL /lokole was not found on this server.
Apache/2.4.29 (Ubuntu) Server at 192.168.0.121 Port 80
```
Suggestions?!

**Comment by veeara282** [comment]
Interesting, it worked before.

**Comment by holta** [comment]
> Interesting, it worked before.

Clarif: `systemctl status lokole.service` confirms this systemd service is indeed `active (running)`.

**Comment by veeara282** [comment]
I'm running `./iiab-install --reinstall` right now. Currently it's getting to `Running local playbooks....Stage 0 will now run....followed by Stages 1-9` and staying there for a long time. After 18 minutes of waiting I stopped the script.

**Comment by veeara282** [comment]
@holta Did you make sure that the role still worked with the changes you suggested?

**Comment by holta** [comment]
> I'm running `./iiab-install --reinstall` right now. Currently it's getting to `Running local playbooks....Stage 0 will now run....followed by Stages 1-9` and staying there for a long time.

`./iiab-install --reinstall` is reliable for me; did you solve the above delay?

The following also works, if you just want to run (the roles in) Stage 6:
```
cd /opt/iiab/iiab
./runrole 6-generic-apps
```
So long as you've included the following in /etc/iiab/local_vars.yml :
```
lokole_install: True
lokole_enabled: True
```
Personally I prefer a surgical strike &mdash; only running the role in question:
```
cd /opt/iiab/iiab
./runrole lokole
```

QUESTION: if [roles/lokole/templates/lokole.conf.j2](https://github.com/aidan-fitz/iiab/blob/lokole-v2/roles/lokole/templates/lokole.conf.j2) needs to be revised (`/var/log/apache2/access.log` shows a number of "GET /lokole HTTP/1.1" 404 errors) &mdash; do [roles/kiwix/templates/kiwix.conf.j2](https://github.com/iiab/iiab/blob/master/roles/kiwix/templates/kiwix.conf.j2) and similar Apache .conf files under [/opt/iiab/iiab/roles](https://github.com/iiab/iiab/tree/master/roles) provide any hints?

**Comment by holta** [comment]
> Did you make sure that the role still worked with the changes you suggested?

I have nothing new to report, and don't believe the overnight syntax errors & tweaks are related in any way.  As outlined [above](https://github.com/iiab/iiab/pull/1282#issuecomment-435612900) (an hour ago) the following does complete:
```
cd /opt/iiab/iiab
git checkout -b aidan-fitz-lokole-v2 master
git pull https://github.com/aidan-fitz/iiab.git lokole-v2
./runrole lokole
```
(But http://box/lokole fails with 404 errors...)

File: roles/lokole/templates/lokole.conf.j2 [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+<Location {{ lokole_url }}/>
+  ProxyPassReverse /
+  ProxyHTMLEnable On
+  ProxyHTMLURLMap / {{ lokole_url }}/
+  RequestHeader unset Accept-Encoding
+</Location>
+
+# /static directory is stored on filesystem
+Alias {{ lokole_url }}/static {{ lokole_install_path }}
```
Comment (holta) [inline roles/lokole/templates/lokole.conf.j2:14]: As is http://box/lokole fails, and http://box/lokole/static shows an imperfect/incomplete page with error:
```
This page does not exist.
```
FWIW when line 14 is shortened down to `Alias {{ lokole_url }} {{ lokole_install_path }}` and then `systemctl reload apache2` is run, it does seem to bypass all (immediate) errors, as a temporary workaround anyway.

i.e. this allowed me to use the http://box/lokole web interface to test creating "email addresses" such as USERNAME@iiab-box.lokole.ca

_(But I don't want to interfere with the intentions of the person who wrote roles/lokole/templates/lokole.conf.j2 &mdash; @aidan-fitz can you perhaps ask them to explain more?)_

File: roles/lokole/templates/lokole.conf.j2 [hunk 10]
```diff
@@ -0,0 +1,27 @@
+# Root directory goes to Lokole web server
+
```
Comment (holta) [inline roles/lokole/templates/lokole.conf.j2:2]:
```suggestion

RedirectMatch ^{{ lokole_url }}$ {{ lokole_url }}/

```

**Review by holta — APPROVED** [review]
@aidan-fitz this would appear to fix it!  What do you think?

**Comment by holta** [comment]
There are of course remaining bugs &mdash; FYI some quick feedback:
- Clicking `Reply` or `Reply all` does not pre-populate the email address(es) into the `To` field _(seems important!)_
- Some emails (including the 1st email sent?) are permanently stuck in the "Outbox" folder (rather than the "Sent" folder) even despite the fact the email was successfully sent _(seems less important)_
- http://box/lokole/static still shows red/pink error `This page does not exist.` _(not sure if this is important?)_

**Comment by holta** [comment]
Suggestions:
- Can you add a warning in [roles/lokole/README.rst](https://github.com/iiab/iiab/blob/master/roles/lokole/README.rst) explaining that you cannot (for now?) email real email addresses?
- Can you document how to freeze (or perhaps just delete) Lokole accounts?  Teachers/libraries/superusers need this when harassment becomes severe (etc) getting in the way of education.
- Can you document how a teacher/librarian/superuser might change an account's password, e.g. when a student/user forgets their password and/or their account password was hacked/changed by someone who should not have?  _(This might also solve the above issue as well ...how to empower teachers/librarians/superusers to "freeze accounts" ...i.e. by changing accounts' passwords?)_
- Should the right-side of the email address (currently appearing as USERNAME@iiab-box.lokole.ca) correspond to {{ iiab_hostname }}.{{ iiab_domain }} or similar?  i.e. USERNAME@box.lan for starters?  Until the implementer further customizes this with some kind of FQDN-like variable we might place in https://github.com/iiab/iiab/blob/master/roles/lokole/defaults/main.yml ?

Ref: https://github.com/ascoderu/opwen-webapp/issues/81

File: roles/lokole/templates/lokole.conf.j2 (hunk 10 above)
Comment (veeara282) [reply roles/lokole/templates/lokole.conf.j2:2]: How would this fix the issue?

Comment (holta) [reply roles/lokole/templates/lokole.conf.j2:2]: > How would this fix the issue?

Magic!  In any case: http://box/lokole now redirects to http://box/lokole/ and the issue is resolved.  As merged into IIAB 6.7/master (PR #1284) and tested on Ubuntu 18.04 and Raspbian Lite 2018-10-09.

**Comment by c-w** [comment]
> Can you document how to freeze (or perhaps just delete) Lokole accounts? Teachers/libraries/superusers need this when harassment becomes severe (etc) getting in the way of education.

@aidan-fitz FYI: Admin accounts can suspend and unsupend non-admin accounts from the `/users` dashboard. Could you add this to the IIAB documentation?

> Can you document how a teacher/librarian/superuser might change an account's password, e.g. when a student/user forgets their password and/or their account password was hacked/changed by someone who should not have?

@aidan-fitz FYI: Admin accounts can change the passwords for non-admin accounts to a random string from the `/users` dashboard. The user can then log in with the temporary password and self-service change it to something more memorable. Could you add this to the IIAB documentation?

Admin accounts can be created via the `./manage.py createadmin` command ([code reference](https://github.com/ascoderu/opwen-webapp/blob/27cb3131e119562f499556a8738d575f40c7cf41/manage.py#L40-L51)). Admins can also promote other user accounts to admin status from the `/users` dashboard.

> Should the right-side of the email address (currently appearing as USERNAME@iiab-box.lokole.ca) correspond to {{ iiab_hostname }}.{{ iiab_domain }} or similar? i.e. USERNAME@box.lan for starters?

@aidan-fitz FYI the email domain can be configured via the `OPWEN_ROOT_DOMAIN` environment variable ([code reference](https://github.com/ascoderu/opwen-webapp/blob/27cb3131e119562f499556a8738d575f40c7cf41/opwen_email_client/webapp/config.py#L13)) and the subdomain can be configured via the `OPWEN_CLIENT_NAME` environment variable ([code reference](https://github.com/ascoderu/opwen-webapp/blob/27cb3131e119562f499556a8738d575f40c7cf41/opwen_email_client/webapp/config.py#L104)). So to achieve the configuration requested above, we can set the following environment variables:

```sh
OPWEN_ROOT_DOMAIN="lan"
OPWEN_CLIENT_NAME="box"
```

**Comment by c-w** [comment]
> http://box/lokole/static still shows red/pink error This page does not exist. (not sure if this is important?)

This is expected. The only files served from `/static` are the JS, CSS, images, etc. and we disable directory listing for the endpoint. When using the app normally, a user should never get into the situation where they navigate to `/static` since it's only used by the HTML to pull in the asset files.

**Comment by holta** [comment]
Many thanks @c-w for the above explanations.  As this PR is now closed, tracking ticket #1293 was just created to bring these to completion &mdash; targeting IIAB 6.7's release in December 2018.

**Pull request merged**